Enable ephemeral compression levels

kdt3rd opened the pull request:
This adds temporary settings to ImfHeader to control the compression levels used.

kdt3rd committed:
Enable deprecating attributes, deprecate dwaCompression

Signed-off-by: Kimball Thurston <kdt3rd@gmail.com>

kdt3rd committed:
Add ephemeral settings for compression levels, propagate to compressors

NB: This changes the ABI of ImfHeader, so must be a 3.2 feature

Signed-off-by: Kimball Thurston <kdt3rd@gmail.com>

kdt3rd commented:
NB: This does change the ABI of the Header class, as well as the internal compression classes (although those matter less) so should not be merged into 3.1.

When reviewing, I did run clang-format on these files.

@aras-p This is the proposal for the temporary settings I mentioned in #1125. I have NOT changed the default value yet. If / once this is accepted, please submit a PR to change to your suggested new default of 4, such that you get the proper credit due for testing all this

kdt3rd committed:
Add ephemeral compression level settings to core

with unit tests

Signed-off-by: Kimball Thurston <kdt3rd@gmail.com>

aras-p commented:
> If / once this is accepted, please submit a PR to change to your suggested new default of 4

Cool, will do!

The formatting of touched files makes this PR look larger than it actually is. On the other hand, looks like you found that e.g. `_readsNothing` used to not be quite initialized, so 🎉

lgritz commented:
This is confusing to read. Can I suggest that you first make a separate PR with just the clang-format changes? Format changes should always stand alone (and their commits can be listed in .git-blame-ignore-revs).

Then on top of that, a separate PR can implement the compression changes and be smaller and easier to read.

Can you clarify the current plan about how to implement the parts of Aras's compression suggestions? I agree that eventually we want to be able to specify compression per-file, but it seems like this solution at least changes ABI, which means it's a minimum of 3.2 (or is it 4.0? will it require changes to client source code?). Because VFX Platform will only allow binary-compatible upgrades during the year, that means the BIG WIN of compression changes won't hit most users until 2023, which seems to me like a real missed opportunity to get a great performance improvement across many applications.

Might we consider a "global" setting for default compression and a standalone function to set it (a little like how thread pool size is done now)? This would not allow the fine-grained per-file control, but it would be binary compatible and thus backportable to a 3.1 patch immediately, and probably is the only granularity of control that most applications need.

Note that this is very different from the dwa compression settings, which must be per-file because they affect image quality as well as size -- there is no one correct setting. But the zip compression level is lossless either way, it's purely a size-vs-time tradeoff, and I suspect that unless there is a problem with Aras's testing methodology, we'll end up changing the default to the new way and the only people who want the control at all will be those who have testing/validation situations where they must produce files that are bit-for-bit identical with saved reference images.

kdt3rd commented:
Yeah, I accidentally hit clang-format in the middle of making the changes. I had thought we had actually run it on the whole repo, but I guess not :( Let me remake w/o that, you're right, it will be cleaner.

kdt3rd commented:
So yes, this does change the ABI, in that the layout for ImfHeader is changing - the changes to the compression objects are internal, and not exported, so would be fine. I guess I am already in the mode that I am going to start changing the c++ ABI fairly dramatically soon, but your point about it then not getting deployed for a while is quite valid...

I did briefly consider hiding the values for the compression in the attribute map, using the empty string as the key, which is an invalid attribute name as far as the rest of the system is concerned, so should never conflict. However, that seemed like it might be "too" clever. That would give us per-part control over this, which might be interesting to see if different AOVs we can actually push further.

So we could do that if you'd prefer, I can write that up and change the above to that to see if you'd prefer that. That would eliminate the external ABI change, and we are only adding non-virtual functions, which would be the same as if we added global functions, so technically are safe to do so to get this in sooner. I don't really like global functions / values because it seems like a bodge, and inconsistent w/ the way the dwaCompression works, which also uses zipCompression for some modes, so it seems like they should be consistently edited, but maybe that's just me

kdt3rd commented:
Actually, strike the empty string in the map - I forgot about the iterators we provide, that would be confusing to people. If we all agree we would like to set it on a per-part basis (I am asking - I think that is what feels right to me), like the dwa compression, we could add that api (not abi breaking), but instead of storing the values in the header, we could store them internally to the library when not default, allowing us to undo the abi breakage so we could put this into 3.1

lgritz commented:
I'm not against any of the flexibility eventually. I just plead for a way to unlock the write speed improvement for 3.1 users in 2022.

Though I definitely see wanting to set the compression *method* -- including the quality level of a lossy codec -- separately per part (that is, per AOV or type of data), I'm not sure that there is a real-world use case for setting the runtime choice of *lossless* compression differently per part. I'm not even confident people will want it different per file. My intuition is that an entire app/studio will make the decision "let's sacrifice 1% disk space for 2x write speed" across the board. But I do know that they want it to be a runtime option, not a build time option, because (a) it makes comparison testing easier to not need separate builds, and (b) many users are stuck with the compiled library that their studio/distro/DCC gives them and don't have the freedom to rebuild from source to enable a big speedup.

kdt3rd committed:
Add layer to avoid changing ABI such that this can be ported to 3.1

Signed-off-by: Kimball Thurston <kdt3rd@gmail.com>

kdt3rd commented:
I've added an internal stash such that the ABI does not have to change to enable this. I ran abi-compliance-checker against the OpenEXR libraries on my system (not compiled by me), and it finds that I've added some functions, but is otherwise clean. It would be good if someone else could validate that as well, although I believe I did the right tests, as when I compare against current master, it correctly identifies the oss fuzz api difference as a problem

kdt3rd committed:
Add global setting of the default compression levels

Signed-off-by: Kimball Thurston <kdt3rd@gmail.com>

kdt3rd commented:
Further to @lgritz request, I have made the global default accessible. This way you can set the global default at runtime as well as compiling it in, but the setting in the header will still allow one to control it on a per-part basis as desired

lgritz commented on cmake/OpenEXRConfig.h.in:
```diff
@@ -147,4 +147,16 @@
 
 #endif
 
+#if defined(__cplusplus) && (__cplusplus >= 201402L)
+# define OPENEXR_DEPRECATED(msg) [[deprecated(msg)]]
+#endif
+
+#ifndef OPENEXR_DEPRECATED
+# ifdef _MSC_VER
+#  define OPENEXR_DEPRECATED(msg) __declspec(deprecated(msg))
+# else
+#  define OPENEXR_DEPRECATED(msg) __attribute__((deprecated(msg)))
+# endif
+#endif
```
Not critical, but would this be simpler as the combined

```
#if defined(__cplusplus) && (__cplusplus >= 201402L)
#  define OPENEXR_DEPRECATED(msg) [[deprecated(msg)]]
#elif defined(_MSC_VER)
#  define OPENEXR_DEPRECATED(msg) __declspec(deprecated(msg))
#else
#  define OPENEXR_DEPRECATED(msg) __attribute__((deprecated(msg)))
#endif
```

lgritz commented on src/lib/OpenEXR/ImfCompression.h:
```diff
@@ -49,6 +49,13 @@ enum IMF_EXPORT_ENUM Compression
 
+/// Controls the default zip compression level used. Zip is used for
+/// the 2 zip levels as well as some modes of the DWAA/B compression.
+IMF_EXPORT void setDefaultZipCompressionLevel (int level);
```
👍

lgritz reviewed (approved):
I can't wait to try this out!

aras-p commented on src/lib/OpenEXR/ImfHeader.cpp:
```diff
@@ -60,8 +62,56 @@ using IMATH_NAMESPACE::Box2i;
+    int zip_level;
+    float dwa_level;
+};
+static std::map<const void *, CompressionRecord> s_globCompressionStore;
```
I never really thought about this, but oh my, the pains/workarounds one has to go through to avoid breaking C++ ABI are sad. One more reason to avoid C++ based public interfaces, I guess. If everything was behind opaque types with a C-like interface (very much like the new OpenEXRCore), none of this would have to be done.

lgritz replied:
Well, to be fair, it's how we did it, not necessarily C vs C++ per se. The original C++ could have used a PIMPL idiom to prevent exposure of the internal data layouts, and a C API can still unwisely expose structure internals.

kdt3rd committed:
Fix issue with static libraries and order of deletion on MacOS

Signed-off-by: Kimball Thurston <kdt3rd@gmail.com>

kdt3rd committed:
add matching functions to control global default zip / dwa levels to C core

Also adds the mechanism to the context initializer

Signed-off-by: Kimball Thurston <kdt3rd@gmail.com>

kdt3rd committed:
Alternate implementation to attempt to fix MacOS and Windows at the same time

Signed-off-by: Kimball Thurston <kdt3rd@gmail.com>

meshula commented on src/lib/OpenEXR/ImfHeader.cpp:
```diff
@@ -60,8 +62,112 @@ using IMATH_NAMESPACE::Box2i;
+    }
+    ~CompressionStash()
+    {
+        s_stash = nullptr;
```
I think this needs to be protected by the mutex, in order to stall destruction of the map in the case that the mutex has already been acquired, for example by an ImfHeader destructor trying to remove itself from the stash.

kdt3rd replied:
well, it's at app static object cleanup time that's the issue, and only when compiling statically (non-deterministic translation unit shutdown ordering) - the first version of this I was getting crashes trying to lock a destroyed mutex, so not sure the extra checks are actually going to help. However, I forgot to check whether threading was enabled or not in the library, so will add those ifdefs and a safety lock / clear just to be a bit safer.

kdt3rd committed:
Add extra safety checks, proper ifdef checks against threading enabled

Also, comment the more complex workaround to get to a stable static
destruction state when compiling to static libraries

Signed-off-by: Kimball Thurston <kdt3rd@gmail.com>

aras-p commented:
@kdt3rd I'm trying to use this in redoing #1125 and some behavior is not quite expected, i.e. 
```
Header h;
h.zipCompressionLevel() = 1;
RgbaOutputFile rgba(myCustomOutStream, h);
```
the final output file does _not_ use the modified compression level. From a quick look it seems that it's because the header gets copied / copy-constructed a few times for the `RgbaOutputFile`. But the Header copy- and move-constructors do not copy over the compression settings record.

Maybe they are just missing a call to `copyCompressionRecord`? In my testing (https://github.com/AcademySoftwareFoundation/openexr/pull/1125/files) adding the copies there does make everything work as expected.

Jamaika1 commented on src/lib/OpenEXR/ImfStandardAttributes.h:
```diff
@@ -45,19 +45,38 @@
 #define IMF_HAS_SUFFIX(suffix) has##suffix
 #define IMF_NAME_ATTRIBUTE(name) name##Attribute
```
```
ImfStandardAttributes.cpp:55:41: warning: 'const OpenEXR_3_1_sidefx::TypedAttribute<float>& OpenEXR_3_1_sidefx::dwaCompressionLevelAttribute(const OpenEXR_3_1_sidefx::Header&)' is deprecated: use compression method in ImfHeader [-Wdeprecated-declarations]
   55 |         return IMF_NAME_ATTRIBUTE(name) (header).value();           \
      |                                         ^
ImfStandardAttributes.cpp:99:1: note: in expansion of macro 'IMF_STD_ATTRIBUTE_IMP'
   99 | IMF_STD_ATTRIBUTE_IMP (dwaCompressionLevel, DwaCompressionLevel, float)
      | ^~~~~~~~~~~~~~~~~~~~~
In file included from ImfStandardAttributes.cpp:12:
ImfStandardAttributes.cpp:99:24: note: declared here
   99 | IMF_STD_ATTRIBUTE_IMP (dwaCompressionLevel, DwaCompressionLevel, float)
      |                        ^~~~~~~~~~~~~~~~~~~
ImfStandardAttributes.h:46:34: note: in definition of macro 'IMF_NAME_ATTRIBUTE'
   46 | #define IMF_NAME_ATTRIBUTE(name) name##Attribute
      |                                  ^~~~
ImfStandardAttributes.cpp:99:1: note: in expansion of macro 'IMF_STD_ATTRIBUTE_IMP'
   99 | IMF_STD_ATTRIBUTE_IMP (dwaCompressionLevel, DwaCompressionLevel, float)
      | ^~~~~~~~~~~~~~~~~~~~~
```
